airbyte-integrations/connectors/source-dixa/source_dixa/source.py

```diff
@@ -0,0 +1,195 @@
+    primary_key = "id"
+    cursor_field = "updated_at"
+
+    def __init__(self, start_timestamp: int, batch_size: int, logger: AirbyteLogger, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.start_timestamp = ConversationExport._validate_ms_timestamp(start_timestamp)
+        # The upper bound is exclusive.
+        self.end_timestamp = ConversationExport.datetime_to_ms_timestamp(datetime.now()) + 1
+        self.batch_size = batch_size
+        self.logger = logger
+
+    @staticmethod
+    def _validate_ms_timestamp(milliseconds: int) -> int:
+        if not type(milliseconds) == int or not len(str(milliseconds)) == 13:
+            raise ValueError(
+                f"Not a millisecond-precision timestamp: {milliseconds}"
+            )
+        return milliseconds
+
+    @staticmethod
+    def ms_timestamp_to_datetime(milliseconds: int) -> datetime:
+        """
+        Converts a millisecond-precision timestamp to a datetime object.
+        """
+        return datetime.fromtimestamp(
+            ConversationExport._validate_ms_timestamp(milliseconds) / 1000
+        )
+
+    @staticmethod
+    def datetime_to_ms_timestamp(dt: datetime) -> int:
+        """
+        Converts a datetime object to a millisecond-precision timestamp.
+        """
+        return int(dt.timestamp() * 1000)
+
+    @staticmethod
+    def add_days_to_ms_timestamp(days: int, milliseconds: int) -> int:
+        return ConversationExport.datetime_to_ms_timestamp(
+            ConversationExport.ms_timestamp_to_datetime(
+                ConversationExport._validate_ms_timestamp(milliseconds)
+            ) + timedelta(days=days)
+        )
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_params(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        self.logger.info(
+            f"Sending request with updated_after={stream_slice['updated_after']} and "
+            f"updated_before={stream_slice['updated_before']}"
+        )
+        return stream_slice
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        yield from response.json()
+
+    def backoff_time(self, response: requests.Response):
+        """
+        The rate limit is 10 requests per minute, so we sleep for one minute
+        once we have reached 10 requests.
+
+        See https://support.dixa.help/en/articles/174-export-conversations-via-api
+        """
+        return 60
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs):
+        """
+        Returns slices of size self.batch_size.
+        """
+        slices = []
+
+        if stream_state and ConversationExport.cursor_field in stream_state:
+            updated_after = stream_state[ConversationExport.cursor_field]
+        else:
+            updated_after = self.start_timestamp
+        updated_before = min(
```

Comment:
would this block be simpler if we do something like: 

```
slices = []
updated_after = ...
while updated_after < self.end_timestamp:
   updated_before = min(updated_after + batch, etc....)
   slices.append({'before':...})
   updated_after = updated_before

return slices
```

This way we're only appending once inside the while loop

Just a small readability improvement probably

Reply:
You're right - I tried to do something like that at first but couldn't get it to work as expected. It seems fine now, and tests are passing so 🤷